File: array_and_hashing/group_anagrams.49.cpp
// O(n*klogk) time
// O(n*k) space
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        std::unordered_map<string, std::vector<string>> groups;
        for (const string& word: strs) {
            // O(k)
            string key = word; // eat
            // O(k*logk)
            sort(key.begin(), key.end()); // aet
            // O(k) for hashing the key + O(1) amortized for push_back
            groups[key].push_back(word); 
        }
        std::vector<std::vector<string>> result;
        for (std::pair<const std::string, std::vector<std::string>>& pair: groups) {
            result.push_back(pair.second); // O(n⋅k) total — copies all strings
        }
        return result;
    }
};

File: array_and_hashing/group_anagrams.49.py
# O(n*klogk) time: k is max word length. For each word, sorted(word) costs O(k log k), done n times.
# O(n*k) space: The hash map stores every word once, total characters across all words is n·k                                   
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        from collections import defaultdict
        groups = defaultdict(list)
        for word in strs:
            key = ''.join(sorted(word)) # built in sorted() for any iterable, returns a new list
            groups[key].append(word) # safe, strings can't be mutated. no need for copy [:]
        return list(groups.values())
    
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = {}
        for word in strs:
            key = ''.join(sorted(word))
            if key not in groups:
                groups[key] = []
            groups[key].append(word)
        return list(groups.values())
    
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = {}
        for word in strs:
            key = ''.join(sorted(word))
            groups.setdefault(key, []).append(word)
        return list(groups.values())